rust/src/applayerpgsql/pgsql.rs

```diff
@@ -0,0 +1,701 @@
+        localstorage_new: None,
+        localstorage_free: None,
+        get_files: None,
+        get_tx_iterator: Some(rs_pgsql_state_get_tx_iterator),
+        get_tx_data: rs_pgsql_get_tx_data,
+        apply_tx_config: None,
+        flags: APP_LAYER_PARSER_OPT_ACCEPT_GAPS,
+        truncate: None,
+    };
+
+    let ip_proto_str = CString::new("tcp").unwrap();
+
+    if AppLayerProtoDetectConfProtoDetectionEnabled(
+        ip_proto_str.as_ptr(),
+        parser.name,
+    ) != 0
+    {
+        let alproto = AppLayerRegisterProtocolDetection(&parser, 1);
+        ALPROTO_POSTGRESQL = alproto;
+        if AppLayerParserConfParserEnabled(
+            ip_proto_str.as_ptr(),
+            parser.name,
+        ) != 0
+        {
+            let _ = AppLayerRegisterParser(&parser, alproto);
+        }
+        SCLogNotice!("Rust PostgreSQL parser registered.");
+    } else {
+        SCLogNotice!("Protocol detector and parser disabled for PostgreSQL.");
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_probe_ssl_request() {
+
+        let buf: &[u8] = &[0x00, 0x00, 0x00, 0x08, 0x04, 0xd2, 0x16, 0x2f];
+        let res = probe_request(&buf);
+        assert!(res.is_ok());
+
+        let buf: &[u8] = &[0x00, 0x00, 0x00, 0x08, 0x04, 0xd2, 0x16, 0x30];
+        let res = probe_request(&buf);
+        assert!(res.is_err());
+
+        let buf: &[u8] = &[0x00, 0x00, 0x00, 0x08, 0x04, 0xd2];
+        let res = probe_request(&buf);
+        assert!(res.is_err());
+        println!("{}", res.unwrap_err());
+    }
+
+    #[test]
+    fn test_probe_ssl_response() {
+        // 'S'
+        let buf: &[u8] = &[0x53];
+        let res = probe_response(&buf);
+        assert!(res.is_ok());
+
+        // 'N'
+        let buf: &[u8] = &[0x4e];
+        let res = probe_response(&buf);
+        assert!(res.is_ok());
+
+        let buf: &[u8] = &[0x30];
+        let res = probe_response(&buf);
+        assert!(res.is_err());
+
+        let buf: &[u8] = &[0x30, 0x31];
+        let res = probe_response(&buf);
+        assert!(res.is_err());
+    }
+
+    #[test]
+    fn test_incomplete_request() {
+        let mut state = PGSQLState::new();
+        let buf: &[u8] = &[0x00, 0x00, 0x00, 0x08, 0x04, 0xd2, 0x16, 0x2f];
+
+        // If we have nothing, we shouldn't be able to parse anything
+        let r = state.parse_request(&buf[0..0]);
+        SCLogNotice!("AppLayerResult is: {:?}", r);
+        assert_eq!(r, AppLayerResult{ status: 0, consumed: 0, needed: 0});
+
+        let r = state.parse_request(&buf[0..1]);
+        SCLogNotice!("AppLayerResult is: {:?}", r);
+        assert_eq!(r, AppLayerResult{ status: 1, consumed: 0, needed: 4});
+
+        let r = state.parse_request(&buf[0..2]);
+        SCLogNotice!("AppLayerResult is: {:?}", r);
+        assert_eq!(r, AppLayerResult{ status: 1, consumed: 0, needed: 4}); // TODO I don't understand why needed is still 4...
```

Comment:
because you (rightfully) didn't consume anything yet.

Reply:
Still wrapping my head around how these things work, I was somehow expecting to see what nom had consumed, that that wouldn't make sense out here, indeed.